src/golang/lib/engine/authenticate.go

```diff
@@ -42,11 +46,40 @@ func AuthenticateLambdaConfig(ctx context.Context, authConf auth.Config) error {
 		lambda_utils.SnowflakeConnectorType,
 	}
 
-	for _, functionType := range functionsToShip {
-		err := lambda_utils.CreateLambdaFunction(functionType, lambdaConf.RoleArn)
-		if err != nil {
-			return errors.Wrap(err, "Unable to Create Lambda Function")
+	errGroup, _ := errgroup.WithContext(ctx)
+
+	// Run authentication only once since the credentials will be memorized.
+	err = lambda_utils.AuthenticateDockerToECR()
+	if err != nil {
+		return errors.Wrap(err, "Unable to Create Lambda Function.")
+	}
+
+	// Pull images on a currency of "MaxConcurrentDownload" to parallelize while avoiding pull timeout.
+	for i := 0; i < len(functionsToShip); i += MaxConcurrentDownload {
+		for j := 0; j < MaxConcurrentDownload; j++ {
```

Comment:
This doubly-nested for-loop is a bit of a brittle construction, and is suboptimal since work is done in strict batches.

I think the canonical way to do this is to use Go channels. Put all the work to do in a single golang channel, to be consumed by some number of errgroup goroutines (workers). Each worker will keep going until 1) the input channel is empty 2) an error occurs or 3) the context is cancelled.